### PR TITLE
Support for Multi-Threaded Execution Plans

### PR DESCRIPTION
**Pull request by jjfumero**
#### Description

This patch enables support for multiple Java threads running different instances of the TornadoVM execution plans. 
The core runtime systems needed to be changed due to not thread-safe when it comes to managing command queues (or CUDA Streams, or Level Zero Command Queues and Command Lists). 

The key part is that there is at least one command queue associated to an execution plan. A execution plan can have more than 1 Java threads associated, if the concurrent-devices or dynamic reconfiguration is also enabled. 

Command Queues are now managed by a double HashMap:

```bash
executionPlanId -> Hash<Device, CommandQueTable> -> Hash<ThreadId, CommandQueue> 
```

The following use case is now allowed:

```java
Thread t0;
Thread t1;

final int size = 1024 * 1024 * 32;
FloatArray input = new FloatArray(size);
input.init(1.0f);
FloatArray output = new FloatArray(size);

TaskGraph taskGraph = new TaskGraph("check") //
       .transferToDevice(DataTransferMode.EVERY_EXECUTION, input) //
       .task("compute01", MultiThreaded::computeForThread2, input, output) //
       .transferToHost(DataTransferMode.EVERY_EXECUTION, output);

// Build two execution plans from two different Java threads that share the same Task-Graph
t0 = new Thread(() -> {
        ImmutableTaskGraph immutableTaskGraph = taskGraph.snapshot();
        TornadoExecutionPlan executionPlan = new TornadoExecutionPlan(immutableTaskGraph);
        executionPlan.execute();
});

t1 = new Thread(() -> {
       ImmutableTaskGraph immutableTaskGraph = taskGraph.snapshot();
       TornadoExecutionPlan executionPlan = new TornadoExecutionPlan(immutableTaskGraph);
       executionPlan.execute();
});

t0.start();
t1.start();

t0.join();
t1.join();
```

#### Backend/s tested

Mark the backends affected by this PR.

- [X] OpenCL
- [X] PTX
- [X] SPIRV

#### OS tested

Mark the OS where this PR is tested.

- [X] Linux
- [ ] OSx
- [ ] Windows

#### Did you check on FPGAs?

If it is applicable, check your changes on FPGAs.

- [ ] Yes
- [X] No

#### How to test the new patch?

```bash
## Pass Unittests using the OpenCL backend
$ make BACKEND=opencl
$ make tests

## Pass unittests using the PTX backend
$ make BACKEND=ptx
$ make tests 

## Pass unittests using the SPIRV backend
$ make BACKEND=spirv
$ make tests 
```

Tests for Multi-threaded:

```bash
$ make 
$ tornado-test --threadInfo -V --fast uk.ac.manchester.tornado.unittests.multithreaded.MultiThreaded#test01
$ tornado-test --threadInfo -V --fast uk.ac.manchester.tornado.unittests.multithreaded.MultiThreaded#test02
```

All supported tests are passing.
Pending for testing:
A) KFusion
B) RayTracer

### CODE REVIEW

**Comment by jjfumero** [comment]
TornadoVM RayTracer works

**Comment by jjfumero** [comment]
KFusion also works

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/ExecutorFrame.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package uk.ac.manchester.tornado.api;
+
+public class ExecutorFrame {
```
Comment (mikepapadim) [inline tornado-api/src/main/java/uk/ac/manchester/tornado/api/ExecutorFrame.java:20]: Can we add javadoc on this new API call?

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/memory/XPUBuffer.java [hunk 2]
```diff
@@ -22,33 +22,33 @@
 import uk.ac.manchester.tornado.api.exceptions.TornadoMemoryException;
 import uk.ac.manchester.tornado.api.exceptions.TornadoOutOfMemoryException;
 
-public interface ObjectBuffer {
+public interface XPUBuffer {
```
Comment (mikepapadim) [inline tornado-api/src/main/java/uk/ac/manchester/tornado/api/memory/XPUBuffer.java:25]: why xpu ? we dont reuse it in opencl?

File: tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/runtime/OCLTornadoDevice.java [hunk 3]
```diff
@@ -81,35 +81,27 @@
...
 import uk.ac.manchester.tornado.runtime.TornadoCoreRuntime;
-import uk.ac.manchester.tornado.runtime.common.DeviceObjectState;
-import uk.ac.manchester.tornado.runtime.common.KernelArgs;
-import uk.ac.manchester.tornado.runtime.common.RuntimeUtilities;
-import uk.ac.manchester.tornado.runtime.common.Tornado;
-import uk.ac.manchester.tornado.runtime.common.TornadoAcceleratorDevice;
-import uk.ac.manchester.tornado.runtime.common.TornadoInstalledCode;
-import uk.ac.manchester.tornado.runtime.common.TornadoOptions;
-import uk.ac.manchester.tornado.runtime.common.TornadoSchedulingStrategy;
+import uk.ac.manchester.tornado.runtime.common.*;
```
Comment (mikepapadim) [inline tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/runtime/OCLTornadoDevice.java:88]: revert * import

File: tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/runtime/OCLTornadoDevice.java [hunk 4]
```diff
@@ -410,10 +402,10 @@ public int[] checkAtomicsForTask(SchedulableTask task, int[] array, int paramInd
     }
 
     @Override
-    public int[] updateAtomicRegionAndObjectState(SchedulableTask task, int[] array, int paramIndex, Object value, DeviceObjectState objectState) {
+    public int[] updateAtomicRegionAndObjectState(SchedulableTask task, int[] array, int paramIndex, Object value, uk.ac.manchester.tornado.runtime.common.DeviceObjectState objectState) {
```
Comment (mikepapadim) [inline tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/runtime/OCLTornadoDevice.java:405]: impory DeviceObjectState

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/api/TestAPI.java [hunk 5]
```diff
@@ -237,7 +237,7 @@ public void testLazyCopyOut() {
 
         // Mark all device memory buffers as free, thus the TornadoVM runtime can reuse
         // device buffers for other execution plans.
-        executionPlan.freeDeviceMemory();
+        //executionPlan.freeDeviceMemory();
```
Comment (mikepapadim) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/api/TestAPI.java:240]: remove comment

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/fails/TestFails.java [hunk 6]
```diff
@@ -65,8 +65,8 @@ public void test01() {
         // =============================================================================
         FloatArray x = new FloatArray(100);
         FloatArray y = new FloatArray(100);
-//        float[] x = new float[100];
-//        float[] y = new float[100];
+        //        float[] x = new float[100];
```
Comment (mikepapadim) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/fails/TestFails.java:68]: comments

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/multithreaded/MultiThreaded.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+/**
+ * How to run?
+ *
+ * <p>
+ * <code>
+ * $ tornado-test -V --fast uk.ac.manchester.tornado.unittests.multithreaded.MultiThreaded
+ * </code>
+ * </p>
+ */
+public class MultiThreaded extends TornadoTestBase {
```
Comment (mikepapadim) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/multithreaded/MultiThreaded.java:44]: can we add some documentation on this programming approach or maybe update docs?

File: tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/tasks/TornadoTaskGraph.java [hunk 8]
```diff
@@ -52,14 +52,7 @@
...
 import jdk.vm.ci.meta.ResolvedJavaMethod;
-import uk.ac.manchester.tornado.api.GridScheduler;
-import uk.ac.manchester.tornado.api.ImmutableTaskGraph;
-import uk.ac.manchester.tornado.api.KernelContext;
-import uk.ac.manchester.tornado.api.Policy;
-import uk.ac.manchester.tornado.api.TaskGraph;
-import uk.ac.manchester.tornado.api.TornadoDriver;
-import uk.ac.manchester.tornado.api.TornadoExecutionPlan;
-import uk.ac.manchester.tornado.api.TornadoTaskGraphInterface;
+import uk.ac.manchester.tornado.api.*;
```
Comment (mikepapadim) [inline tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/tasks/TornadoTaskGraph.java:55]: * imprort

File: tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/tasks/TornadoTaskGraph.java [hunk 9]
```diff
@@ -225,7 +221,7 @@ static void performStreamInObject(TaskGraph task, List<Object> inputObjects, fin
             case 0:
```
Comment (mikepapadim) [inline tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/tasks/TornadoTaskGraph.java:221]: can we do it with enhcnaed switch to simplify?

File: tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/tasks/LocalObjectState.java [hunk 10]
```diff
@@ -23,26 +23,56 @@
...
+     * For each variable, we need to keep track of all devices in which there is a shadow
+     * copy. This is achieved by using the {@link DataObjectState} object.
+     */
+    private DataObjectState globalObjectState;
+
+    private Object object;
 
     public LocalObjectState(Object object) {
-        global = getTornadoRuntime().resolveObject(object);
+        //globalObjectState = getTornadoRuntime().resolveObject(object);
```
Comment (mairooni) [inline tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/tasks/LocalObjectState.java:65]: Remove comment

File: tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/interpreter/TornadoVMInterpreter.java [hunk 11]
```diff
@@ -166,12 +165,11 @@ public void setTimeProfiler(TornadoProfiler tornadoProfiler) {
     }
 
     public void fetchGlobalStates() {
-        debug("fetching %d object states...", globalStates.length);
         for (int i = 0; i < objects.size(); i++) {
             final Object object = objects.get(i);
             TornadoInternalError.guarantee(object != null, "null object found in TornadoVM");
-            globalStates[i] = TornadoCoreRuntime.getTornadoRuntime().resolveObject(object);
-            debug("\tobject[%d]: [0x%x] %s %s", i, object.hashCode(), object.getClass().getTypeName(), globalStates[i]);
+            //globalStates[i] = TornadoCoreRuntime.getTornadoRuntime().resolveObject(object);
```
Comment (mairooni) [inline tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/interpreter/TornadoVMInterpreter.java:171]: Remove comment

File: tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/runtime/SPIRVTornadoDevice.java [hunk 12]
```diff
@@ -386,92 +377,96 @@ public int deallocate(TornadoDeviceObjectState state) {
...
     }
 
     @Override
-    public void ensureLoaded() {
+    public void ensureLoaded(long executionPlanId) {
+        // create command queue
 
+        // TODO: pass execution plan ID to map EP -> command queues
+        // TODO: This must be managed by a common context
+        // getDeviceContext().getSpirvContext().createCommandQueue(deviceIndex);
```
Comment (mairooni) [inline tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/runtime/SPIRVTornadoDevice.java:433]: Remove comment

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/memory/XPUBuffer.java (hunk 2 above)
Comment (jjfumero) [reply tornado-api/src/main/java/uk/ac/manchester/tornado/api/memory/XPUBuffer.java:25]: `ObjectBuffer` is too broad term. I wanted to refer to a Buffer on the accelerator. This interface is common for all backends, including OpenCL

**Comment by jjfumero** [comment]
All comments addressed

**Review by mairooni — APPROVED** [review]

**Review by mikepapadim — APPROVED** [review]

**Pull request merged**